app/src/main/java/org/wikipedia/categories/db/CategoryDao.kt

```diff
@@ -13,16 +13,31 @@ interface CategoryDao {
     @Insert(onConflict = OnConflictStrategy.REPLACE)
     suspend fun insertAll(categories: List<Category>): List<Long>
 
-    @Query("SELECT title, lang, COUNT(*) as count FROM Category where timeStamp >= :startTimeStamp AND timeStamp <= :endTimeStamp " +
-            "GROUP BY title, lang ORDER BY count DESC")
-    suspend fun getCategoriesByTimeRange(startTimeStamp: Long, endTimeStamp: Long): List<CategoryCount>
+    // find categories by primary keys
+    @Query("SELECT * FROM Category WHERE year = :year AND month = :month AND title = :title AND lang = :lang")
+    suspend fun findByPrimaryKey(year: Int, month: Int, title: String, lang: String): Category?
+
+    // Update count by primary keys
+    @Query("UPDATE Category SET count = count + 1 WHERE year = :year AND month = :month AND title = :title AND lang = :lang")
+    suspend fun updateCountByPrimaryKeys(year: Int, month: Int, title: String, lang: String)
+
+    @Query("SELECT year, month, title, lang, SUM (count) AS count FROM Category WHERE year BETWEEN :startYear AND :endYear GROUP BY title, lang ORDER BY count DESC")
+    suspend fun getCategoriesByTimeRange(startYear: Int, endYear: Int): List<Category>
 
     @Query("SELECT * FROM Category")
     suspend fun getAllCategories(): List<Category>
 
     @Query("DELETE FROM Category")
     suspend fun deleteAll()
 
-    @Query("DELETE FROM Category WHERE rowid IN (SELECT rowid FROM Category WHERE timeStamp <:timeStamp LIMIT :batchSize)")
-    suspend fun deleteOlderThanInBatch(timeStamp: Long, batchSize: Int): Int
+    @Query("DELETE FROM Category WHERE rowid IN (SELECT rowid FROM Category WHERE year <:year LIMIT :batchSize)")
+    suspend fun deleteOlderThanInBatch(year: Int, batchSize: Int): Int
+
+    suspend fun upsertAll(list: List<Category>) {
```

Comment:
Since this function performs multiple db operations, it can be annotated with `@Transaction`, to indicate to Room that this should be executed in a single transaction, which will be more efficient.

Reply:
Done!